File: src/main/java/com/github/simy4/poc/freemarker/TemplateTool.java
package com.github.simy4.poc.freemarker;

import freemarker.ext.beans.StringModel;
import freemarker.template.Template;
import freemarker.template.TemplateModelException;
import org.springframework.stereotype.Service;
import org.springframework.util.ReflectionUtils;

import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

import static java.util.Objects.requireNonNull;

/** Freemarker template utils. */
@Service("freemarkerTemplateTool")
public class TemplateTool {

  /**
   * Retrieves a list of arguments referenced in given template.
   *
   * @param template template to introspect
   * @return set of arguments
   */
  @SuppressWarnings("deprecation")
  public Set<String> referenceSet(Template template) throws TemplateModelException {
    var result = new HashSet<String>();
    var rootTreeNode = template.getRootTreeNode();
    for (var i = 0; i < rootTreeNode.getChildCount(); i++) {
      var templateModel = rootTreeNode.getChildNodes().get(i);
      if (templateModel instanceof StringModel stringModel) {
        processStringModel(stringModel).ifPresent(result::add);
      }
    }
    return result;
  }

  private Optional<String> processStringModel(StringModel stringModel) {
    var wrappedObject = stringModel.getWrappedObject();
    return switch (wrappedObject.getClass().getSimpleName()) {
      case "DollarVariable" -> processDollarVariable(wrappedObject);
      default -> Optional.empty();
    };
  }

  private Optional<String> processDollarVariable(Object dollarVariable) {
    var expression = getInternalState(dollarVariable, "expression");
    return switch (expression.getClass().getSimpleName()) {
      case "Identifier" -> Optional.of(getInternalState(expression, "name").toString());
      case "DefaultToExpression" -> Optional.of(getInternalState(expression, "lho").toString());
      case "MethodCall", "BuiltinVariable" -> Optional.empty();
      default -> throw new IllegalStateException(
          "Unable to introspect variable " + expression + " of type " + expression.getClass());
    };
  }

  private Object getInternalState(Object o, String fieldName) {
    var field = ReflectionUtils.findField(o.getClass(), fieldName);
    ReflectionUtils.makeAccessible(requireNonNull(field, "Field is null"));
    return ReflectionUtils.getField(field, o);
  }
}

File: src/main/kotlin/com/github/simy4/poc/freemarker/TemplateTool.kt
package com.github.simy4.poc.freemarker

import freemarker.ext.beans.StringModel
import freemarker.template.Template
import freemarker.template.TemplateModelException
import org.springframework.stereotype.Service
import org.springframework.util.ReflectionUtils

/** Freemarker template utils. */
@Service("freemarkerTemplateTool")
class TemplateTool {
  @Suppress("DEPRECATION")
  @Throws(TemplateModelException::class)
  fun referenceSet(template: Template): Set<String> =
      template.rootTreeNode.let { rootTreeNode ->
        (0..rootTreeNode.childCount)
            .mapNotNull { i ->
              rootTreeNode.childNodes[i].let { templateModel ->
                when (templateModel) {
                  is StringModel -> processStringModel(templateModel)
                  else -> null
                }
              }
            }
            .toSet()
      }

  private fun processStringModel(stringModel: StringModel): String? =
      stringModel.wrappedObject.let { wrappedObject ->
        when (wrappedObject.javaClass.simpleName) {
          "DollarVariable" -> processDollarVariable(wrappedObject)
          else -> null
        }
      }

  private fun processDollarVariable(dollarVariable: Any): String? =
      dollarVariable.internalState("expression").let { expression ->
        when (expression!!.javaClass.simpleName) {
          "Identifier" -> expression.internalState("name").toString()
          "DefaultToExpression" -> expression.internalState("lho").toString()
          "MethodCall",
          "BuiltinVariable" -> null
          else ->
              throw IllegalStateException(
                  "Unable to introspect variable " +
                      expression +
                      " of type " +
                      expression.javaClass)
        }
      }

  private fun Any.internalState(fieldName: String): Any? {
    val field = ReflectionUtils.findField(javaClass, fieldName)!!
    ReflectionUtils.makeAccessible(field)
    return ReflectionUtils.getField(field, this)
  }
}
